router/usecase/pools/routable_cw_orderbook_pool.go

```diff
@@ -78,20 +80,29 @@ func (r *routableOrderbookPoolImpl) CalculateTokenOutByTokenIn(ctx context.Conte
 
 	// Get starting tick index for the "out" side of the orderbook
 	// Since the order will get the liquidity out from that side
-	tickIdx, err := r.GetStartTickIndex(directionOut)
+	tickIdx, err := r.OrderbookData.GetStartTickIndex(directionOut)
 	if err != nil {
 		return sdk.Coin{}, err
 	}
 
 	amountOutTotal := osmomath.ZeroBigDec()
 	amountInRemaining := osmomath.BigDecFromSDKInt(tokenIn.Amount)
 
+	amountInToExhaustLiquidity := osmomath.BigDec{}
+	if *directionIn == cosmwasmpool.BID {
+		amountInToExhaustLiquidity = r.OrderbookData.BidAmountToExhaustAskLiquidity
+	} else {
+		amountInToExhaustLiquidity = r.OrderbookData.AskAmountToExhaustBidLiquidity
+	}
+
+	// check if amount in > amountInToExhaustLiquidity, if so this swap is not possible due to insufficient liquidity
+	if amountInRemaining.GT(amountInToExhaustLiquidity) {
+		return sdk.Coin{}, domain.OrderbookNotEnoughLiquidityToCompleteSwapError{PoolId: r.GetId(), AmountIn: tokenIn}
+	}
+
 	// ASSUMPTION: Ticks are ordered
 	for amountInRemaining.GT(zeroBigDec) {
-		// Order has run out of ticks to iterate
-		if tickIdx >= len(r.OrderbookData.Ticks) || tickIdx < 0 {
-			return sdk.Coin{}, domain.OrderbookNotEnoughLiquidityToCompleteSwapError{PoolId: r.GetId(), AmountIn: tokenIn}
-		}
```

Comment:
Should we still keep this check as defense-in-depth?

Reply:
sure thing